### PR TITLE
Enable FF to drive API Execution

### PR DESCRIPTION
**Pull request by nigurr**

### CODE REVIEW

File: Tasks/VsTestV2/runvstest.ts [hunk 1]
```diff
@@ -27,38 +29,70 @@ if (osPlat !== 'win32') {
...
             const test = new distributedTest.DistributedTest(inputDataContract);
             test.runDistributedTest();
         } else {
             ci.publishEvent({ runmode: 'nondistributed' });
             console.log(tl.loc('nonDistributedTestWorkflow'));
             console.log('======================================================');
             const inputDataContract = inputParser.parseInputsForNonDistributedTestRun();
+            if (enableApiExecution || (inputDataContract.ExecutionSettings
+                && inputDataContract.ExecutionSettings.RerunSettings
+                && inputDataContract.ExecutionSettings.RerunSettings.RerunFailedTests)) {
```
Comment (acesiddhu) [inline Tasks/VsTestV2/runvstest.ts]: lets refactor this and make this condition simpler

File: Tasks/VsTestV2/runvstest.ts [hunk 2]
```diff
@@ -27,38 +29,70 @@ if (osPlat !== 'win32') {
...
+    finally {
         taskProps.state = 'completed';
         ci.publishEvent(taskProps);
     }
 }
 
+function getFeatureFlag(collectionUri: string, featureFlag: string, token: string): Promise<boolean> {
+    let state = false;
+    const options = {
+        url: collectionUri + '/_apis/FeatureFlags/' + featureFlag,
```
Comment (acesiddhu) [inline Tasks/VsTestV2/runvstest.ts]: >FeatureFlags [](start = 37, length = 12)

general question. are these apis documented?

File: Tasks/VsTestV2/runvstest.ts [hunk 3]
```diff
@@ -27,38 +29,70 @@ if (osPlat !== 'win32') {
...
+        headers: {
+            'Content-Type': 'application/json',
+            'Authorization': 'Bearer ' + token
+        }
+    };
+
+    return new Promise((resolve, reject) => {
+        request(options, (err, res, faModel) => {
+            if (err) {
+                console.warn('Unable to get feature flag ' + featureFlag + ' Error:' + err.message);
```
Comment (acesiddhu) [inline Tasks/VsTestV2/runvstest.ts]: >Unable to get feature flag [](start = 30, length = 26)

localize

File: Tasks/VsTestV2/runvstest.ts [hunk 4]
```diff
@@ -7,18 +7,20 @@ import * as ci from './cieventlogger';
...
-if (osPlat !== 'win32') {
-    // Fail the task if os is not windows
-    tl.setResult(tl.TaskResult.Failed, tl.loc('OnlyWindowsOsSupported'));
-} else {
-    //Starting the VsTest execution
+async function execute() {
     const taskProps = { state: 'started', result: '' };
     ci.publishEvent(taskProps);
 
+    const enableApiExecution = await getFeatureFlag(tl.getVariable('System.TeamFoundationCollectionUri'),
```
Comment (acesiddhu) [inline Tasks/VsTestV2/runvstest.ts]: >getFeatureFlag [](start = 37, length = 14)

rename it to isFeatureFlagEnabled.

getFF should ideally return on and off

File: Tasks/VsTestV2/runvstest.ts (hunk 2 above)
Comment (acesiddhu) [reply Tasks/VsTestV2/runvstest.ts]: answer is yes

---
In reply to: [202637953](https://github.com/Microsoft/vsts-tasks/pull/7669#discussion_r202637953) [](ancestors = 202637953)

File: Tasks/VsTestV2/runvstest.ts (hunk 4 above)
Comment (nigurr) [reply Tasks/VsTestV2/runvstest.ts]: done

**Pull request merged**